lib/spack/spack/build_environment.py

```diff
@@ -792,43 +795,45 @@ def load_external_modules(pkg):
 
 def setup_package(pkg, dirty, context='build'):
     """Execute all environment setup routines."""
-    env = EnvironmentModifications()
+    if dirty:
+        env = EnvironmentModifications()
+    else:
+        env = clean_environment()
 
-    if not dirty:
-        clean_environment()
+    # Keep track of env changes on top of a clean environment, since we want
+    # to validate these for "suspicious" modifications of variables.
+    env_mods = EnvironmentModifications()
 
     # setup compilers and build tools for build contexts
     need_compiler = context == 'build' or (context == 'test' and
                                            pkg.test_requires_compiler)
     if need_compiler:
-        set_compiler_environment_variables(pkg, env)
-        set_build_environment_variables(pkg, env, dirty)
+        env_mods.extend(set_compiler_environment_variables(pkg))
+        env_mods.extend(set_build_environment_variables(pkg, dirty))
 
     # architecture specific setup
-    pkg.architecture.platform.setup_platform_environment(pkg, env)
+    pkg.architecture.platform.setup_platform_environment(pkg, env_mods)
 
     if context == 'build':
         # recursive post-order dependency information
-        env.extend(
-            modifications_from_dependencies(pkg.spec, context=context)
-        )
+        env_mods.extend(modifications_from_dependencies(pkg.spec, context=context))
 
-        if (not dirty) and (not env.is_unset('CPATH')):
+        if (not dirty) and (not env_mods.is_unset('CPATH')):
             tty.debug("A dependency has updated CPATH, this may lead pkg-"
```

Comment:
should this be debug or perhaps a warning? I'm just thinking probably most people don't run things in debug. I know it was already here, but since it's part of the PR it's worth bringing up for discussion.